### PR TITLE
Challenge3 2april

### PR DESCRIPTION
**Pull request by nobeeakon**
**Acceptance Criteria**
- [X] Videos in the Home View are fetched from the YouTube API and displayed correctly.
- [X] Users can search for YouTube videos using the search field on the header.
- [X] A video can be played from the Video Details View after clicking on it.
- [x] The video information and related videos list are displayed correctly on the Video Details View.
- [X] When a user clicks on a related video the video player, information and related videos list are updated.
**Bonus Points**
- [X] Custom Hooks for API calls are implemented and tested correctly.
- [X] Testing coverage is above 60%. (Please include a screenshot of the code coverage report in your PR description).

imagen

### CODE REVIEW

File: src/components/RelatedVideoList/RelatedVideoCard/RelatedVideoCard.spec.jsx [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+          </DarkModeProvider>
+        </BrowserRouter>
+      );
+
+      const videoCard = getByText(/title/i);
+      fireEvent.click(videoCard);
+      expect(mockedGoToVideo).toHaveBeenCalledTimes(1);
+    });
+  });
+});
```
Comment (vmcruz) [inline src/components/RelatedVideoList/RelatedVideoCard/RelatedVideoCard.spec.jsx:129]: Nice testing. An edge case I suggest you to test is what happens with your default props. Try testing the component with only the required properties and see if it behaves as expected. Perhaps you'll find that you need to require another one.

File: src/components/RelatedVideoList/RelatedVideos/RelatedVideos.spec.jsx [hunk 2]
```diff
@@ -105,16 +112,21 @@ describe('Testing VideoList', () => {
...
+            <ThemesProvider>
+              <RelatedVideos relatedToVideoId={relatedToVideoId} />
+            </ThemesProvider>
+          </DarkProvider>
+        </BrowserRouter>
       );
+
+      expect(useVideos).toBeCalledWith(relatedToVideoId, REQ_TYPE);
+      expect(getByText(/Wizeline title1/i)).toBeInTheDocument();
+      expect(getByText(/Video Tour | Welcome /i)).toBeInTheDocument();
```
Comment (vmcruz) [inline src/components/RelatedVideoList/RelatedVideos/RelatedVideos.spec.jsx]: Nice tests. I like that you mocked your hook results, this way you have control over what gets rendered in your tests 👍🏽 .

File: src/components/VideoDetail/VideoDetailView/VideoDetailView.component.jsx [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+                  </Tag>
+                ))}
+            {tags.length > MAX_TAGS && (
+              <ExtraTagsButton
+                onClick={toogleShowExtraTags}
+                data-testid="toggleShowExtraTags"
+              >
+                {!showExtraTags ? 'more...' : 'less...'}
+              </ExtraTagsButton>
+            )}
```
Comment (vmcruz) [inline src/components/VideoDetail/VideoDetailView/VideoDetailView.component.jsx]: This logic is getting harder to understand. Try moving it to a different component, or prepare it outside the `return`, so it is cleaner and easier to understand.

File: src/components/VideoList/VideoList/VideoList.component.jsx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  return (
+    <div>
+      {videoList.length === 0 ? (
+        <NoVideoFound> No Video Found :( </NoVideoFound>
+      ) : (
+        <VideoContainer>
+          {videoList.map((video) =>
+            video.id.videoId ? (
+              <VideoListCard
+                goToVideoHandler={(e) => goToVideo(e, video.id.videoId)}
```
Comment (vmcruz) [inline src/components/VideoList/VideoList/VideoList.component.jsx]: What if you move the `goToVideo` to the VideoListCard itself?, you just need the `videoId`. This way you don't create external dependencies for your component, and you can control the click behavior on the actual component.

File: src/components/VideoList/VideoList/VideoList.spec.jsx [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    it('should display "No Video Found" when result is empty array and isLoading is false', async () => {
+      useVideos.mockReturnValueOnce({ videoList: [], isLoading: false });
+
+      const { container } = render(<VideoList />);
+
+      expect(container).toHaveTextContent(/no video found/i);
+    });
+
+    it('should display video Cards when array is returned', async () => {
+      const returnedArray = [
```
Comment (vmcruz) [inline src/components/VideoList/VideoList/VideoList.spec.jsx:65]: Since this is a quite large mock, I recommend you to move it to a separated `__mocks__/youtubeResponse.js` file (suggested name), so that your tests are cleaner.

File: src/utils/constants.js [hunk 6]
```diff
@@ -1,5 +1,11 @@
...
+const DARK_MODE_STORAGE_KEY = 'is_dark_mode_on';
 
-export { AUTH_STORAGE_KEY, DARK_MODE_KEY };
+const REQUEST_API_TYPES = {
+  SEARCH_BY_TERM: 'search_by_term',
+  SEARCH_RELATED_VIDEOS: 'search_related_videos',
+  VIDEO_INFO: 'video_info',
+};
+
+export { AUTH_STORAGE_KEY, DARK_MODE_STORAGE_KEY, REQUEST_API_TYPES };
```
Comment (vmcruz) [inline src/utils/constants.js:11]: Nice use of a `constants` file.

File: src/utils/functions/fetchData.js [hunk 7]
```diff
@@ -1,8 +1,58 @@
...
+export const fetchVideoInfo = async (videoId) => {
+  const response = await youtubeAPI.get('/videos', {
+    params: {
+      part: ['snippet', 'contentDetails', 'statistics'].join(', '),
+      id: videoId,
+    },
+  });
+
+  return response.data.items;
+};
```
Comment (vmcruz) [inline src/utils/functions/fetchData.js:73]: I like this abstraction layer. Just consider adding error handling for the responses, you could get an exceeded quota, or 500 if something went wrong, or a denied if the api key is invalid.

File: src/utils/hooks/useVideos.js [hunk 8]
```diff
@@ -1,30 +1,53 @@
...
+  }, [requestType, searchedString]);
 
-    fetchVideos();
-  }, []);
+  useEffect(() => {
+    fetchData();
+  }, [fetchData]);
 
-  return [videoList, isLoading];
+  return { videoList, isLoading };
```
Comment (vmcruz) [inline src/utils/hooks/useVideos.js:49]: Good work with this hook, clean and gets the job done.

File: src/utils/hooks/useVideos.spec.js [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    const REQ_TYPE = 'other';
+    const VIDEO_ID = 'Ks-_Mh1QhMc';
+
+    test('After invalid fetching type, videoList =  null ,  isLoading = false', () => {
+      const { result } = renderHook(() => useVideos(VIDEO_ID, REQ_TYPE));
+      expect(result.current.videoList).toBeNull();
+      expect(result.current.isLoading).toBe(false);
+    });
+  });
+});
```
Comment (vmcruz) [inline src/utils/hooks/useVideos.spec.js:116]: I must say that I wasn't expecting a complete setup for the API testing. This is nice!. The only `but` I have to say is that for this scenario is perhaps an overkill. I don't know much about the `msw` package, but it looks like `supertest` perhaps. It would be easier if you just have mocked the `axios` package, or the `.get` implementation. This way you don't add a lot of boilerplate to your project for testing.

**Review by vmcruz — COMMENTED** [review]
I really like your project. It is well written. I left a couple of comments. Good work!.

**Comment by nobeeakon** [comment]
updated branch following to Victor's advices
